### PR TITLE
Qemu : Enable SmartFS

### PR DESCRIPTION
**Pull request by manoharahk**

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/tiva/tiva_flash.c [hunk 1]
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
...
+#if !defined(CONFIG_QEMU_SRAM) &&  !defined (CONFIG_QEMU_SDRAM)
 	FAR uint32_t *src = (uint32_t *) buf;
 	FAR uint32_t *dst = (uint32_t *)(TIVA_VIRTUAL_BASE + startblock * TIVA_FLASH_PAGESIZE);
 	int i;
+#endif
 
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
+	/* Qemu doesnt implement tive flash memory controller.
```
Comment (pradeepns85) [inline os/arch/arm/src/tiva/tiva_flash.c]: tive -> tiva

File: os/arch/arm/src/tiva/tiva_flash.c [hunk 2]
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
...
+
+#if !defined(CONFIG_QEMU_SRAM) &&  !defined (CONFIG_QEMU_SDRAM)
 	FAR uint32_t *src = (uint32_t *) buf;
 	FAR uint32_t *dst = (uint32_t *)(TIVA_VIRTUAL_BASE + startblock * TIVA_FLASH_PAGESIZE);
 	int i;
+#endif
 
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
```
Comment (pradeepns85) [inline os/arch/arm/src/tiva/tiva_flash.c]: check space after defined

File: os/arch/arm/src/tiva/tiva_flash.c [hunk 3]
```diff
@@ -197,12 +208,23 @@ static ssize_t tiva_bread(FAR struct mtd_dev_s *dev, off_t startblock, size_t nb
...
+	/* Qemu doesnt implement tive flash memory controller.
+	 * Hence copy the contents from buf to flash blocks using memcpy.
+	 */
+	memcpy((uint32_t *)(TIVA_VIRTUAL_BASE + startblock * TIVA_FLASH_PAGESIZE), (uint32_t *)buf, nblocks * TIVA_FLASH_PAGESIZE);
```
Comment (pradeepns85) [inline os/arch/arm/src/tiva/tiva_flash.c]: You could have used src and dest variables declared above . Is there any reason for binding them under not defined flag and not using it inside memcpy ?  May be am wrong. just curious to know

File: os/arch/arm/src/tiva/tiva_flash.c [hunk 4]
```diff
@@ -144,11 +144,21 @@ static struct tiva_dev_s g_lmdev = {
...
 {
+
+#if !defined(CONFIG_QEMU_SRAM) &&  !defined (CONFIG_QEMU_SDRAM)
 	int curpage;
 	uint32_t pageaddr;
+#endif
 
 	DEBUGASSERT(nblocks <= TIVA_VIRTUAL_NPAGES);
 
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
```
Comment (pradeepns85) [inline os/arch/arm/src/tiva/tiva_flash.c]: check space after defined ..

File: os/arch/arm/src/tiva/chip/lm3s_flash.h [hunk 5]
```diff
@@ -70,11 +70,19 @@
 	defined(CONFIG_ARCH_CHIP_TM4C123GH6ZRB) || defined(CONFIG_ARCH_CHIP_TM4C123GH6PMI) || \
 	defined(CONFIG_ARCH_CHIP_CC3200)
 
+
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
+/* These parts all support a 1KiB erase page size and a total FLASH memory size
+ * of 128MB or 128K pages.
+ */
+#define TIVA_FLASH_NPAGES        (128*1024)
```
Comment (pradeepns85) [inline os/arch/arm/src/tiva/chip/lm3s_flash.h]: Isn't possible to make it dynamic. for ex. if FLASH_SIZE = 128 * 1024 * 1024 ,   then 
#define TIVA_FLASH_NPAGES  FLASH_SIZE/1024    . This way we don't have to modify this for different memory configurations. what's your opinion ?

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 6]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+
+	if (mtd->ioctl(mtd, MTDIOC_GEOMETRY, (unsigned long)&geo) < 0) {
+		lldbg("ERROR: mtd->ioctl failed\n");
+		return 0;
+	}
+
+#ifdef CONFIG_MTD_PARTITION
+	mtd_part = mtd_partition(mtd, partoffset, partsize / geo.blocksize, partno);
+	if (!mtd_part) {
+		lldbg("ERROR: failed to create partition.\n");
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: may be you can print like
failed to create  mtd partition and if possible, print the reason like error value?

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 7]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+	int minorno = QEMU_SMARTFS_PARTITION_MINORNO;
+	int partoffset = QEMU_SMARTFS_PARTITION_START;
+	int partsize = QEMU_SMARTFS_PARTITION_SIZE;
+
+#ifdef CONFIG_MTD
+	mtd = up_flashinitialize();
+
+	if (!mtd) {
+		lldbg("ERROR : up_flashinitializ failed\n");
+		return 0;
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: returning 0 may be for success case , how about returning -1 for failure cases ?

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 8]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+	}
+
+	if (mtd->ioctl(mtd, MTDIOC_GEOMETRY, (unsigned long)&geo) < 0) {
+		lldbg("ERROR: mtd->ioctl failed\n");
+		return 0;
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: same as above

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 9]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+#ifdef CONFIG_MTD_PARTITION
+	mtd_part = mtd_partition(mtd, partoffset, partsize / geo.blocksize, partno);
+	if (!mtd_part) {
+		lldbg("ERROR: failed to create partition.\n");
+		return 0;
+	}
+#endif /* CONFIG_MTD_PARTITION */
+
+	snprintf(partref, sizeof(partref), "p%d", partno);
+	snprintf(devname, sizeof(devname), "/dev/smart%d%s", minorno, partref);
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: Don't we need to check the return value of snprintf ?

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 10]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+#if defined(CONFIG_MTD_SMART) && defined(CONFIG_FS_SMARTFS)
+	smart_initialize(minorno, mtd_part, partref);
+#endif /* defined(CONFIG_MTD_SMART) && defined(CONFIG_FS_SMARTFS) */
+
+#endif /* CONFIG_MTD */
+
+	ret = mksmartfs(devname, false);
+
+	if (ret != OK) {
+		lldbg("ERROR: mksmartfs on %s failed\n", devname);
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: how about printing the error no?

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 11]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
+	} else {
+		ret = mount(devname, QEMU_SMARTFS_MOUNT_POINT, "smartfs", 0, NULL);
+		if (ret != OK) {
+			lldbg("ERROR: mounting '%s' failed\n", devname);
```
Comment (pradeepns85) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: same as above

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 7 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: Ok. I will change it to return ERROR.
But, return value is not handled at the caller end.

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 8 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: ok

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 6 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: error value is not returned here

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 9 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: It is good to check

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 10 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: yes, we can print ret value too.

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 11 above)
Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: ok

File: os/arch/arm/src/tiva/chip/lm3s_flash.h (hunk 5 above)
Comment (manoharahk) [reply os/arch/arm/src/tiva/chip/lm3s_flash.h]: In my opionion, It just adds one more macro for total flash size then update the existing npage macro. If flash size changes, we need modify flash size macro instead of npage macro.

File: os/arch/arm/src/tiva/tiva_flash.c (hunk 4 above)
Comment (manoharahk) [reply os/arch/arm/src/tiva/tiva_flash.c]: I will check

File: os/arch/arm/src/tiva/tiva_flash.c (hunk 2 above)
Comment (manoharahk) [reply os/arch/arm/src/tiva/tiva_flash.c]: ok

File: os/arch/arm/src/tiva/tiva_flash.c (hunk 1 above)
Comment (manoharahk) [reply os/arch/arm/src/tiva/tiva_flash.c]: Okay, i will correct it

**Comment by pillip8282** [comment]
@manoharahk 
Could you let me know how to apply patch?
Give me a guide

**Comment by pillip8282** [comment]
@manoharahk 
it is not working... it seems that intiailizing smartfs fail.
belows are log when qemu starts
----------------------------------------------------------------
LM3S6965EVB : FLASH (B) = 134217728 : SRAM (B) = 1048576 : SDRAM (B) = 536870912
ABCDF
smart_findfreephyssector: Program bug!  Expected a free sector 1
board_initialize: ERROR: mksmartfs on /dev/smart0p0 failed
System Information:
	Version: 1.1
	Commit Hash: f2a3514a65a6334c8e9468dcaa05a05fcd6caddf
	Build User: kang@kang-linux
	Build Time: 2018-02-01 08:26:25
	System Time: 08 Jun 2010, 00:00:02 [s] UTC 
TASH>>
----------------------------------------------------------------

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c [hunk 12]
```diff
@@ -107,3 +112,86 @@ void tiva_boardinitialize(void)
...
 #endif
 }
+
+
+#ifdef CONFIG_BOARD_INITIALIZE
+
+#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)
+
+#define 	QEMU_SMARTFS_PARTITION_START	(0)
+#define 	QEMU_SMARTFS_PARTITION_SIZE	(1024 * 1024) /* 1MB */
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: please check here, tiva using below as a whole flash size.
 #define TIVA_FLASH_NPAGES        256
 #define TIVA_FLASH_PAGESIZE      1024

**Comment by manoharahk** [comment]
Change Log:
1. Added a patch to update qemu 16M patch for supporting read/write flash 
2. Error handling for return case
3. snprintf error handling
4. Fixed space issue,
5. Updated memcpy args during tiva_write

**Comment by manoharahk** [comment]
@pillip8282 ,

I have updated the  qemu-2.10.0-rc2_16m_ram_size.patch.
Please apply the qemu-2.10.0-rc2_16m_ram_size.patch on qemu-2.10.0-rc2 , build and check.

Steps to build qemu (As documented in build/configs/qemu/README.md):
wget http://download.qemu-project.org/qemu-2.10.0-rc2.tar.xz                     
tar xvJf qemu-2.10.0-rc2.tar.xz                                                  
cd qemu-2.10.0-rc2                                                                                                  
/* If you want to use 16MB RAM, copy qemu-2.10.0-rc2_16m_ram_size.patch from TizenRT/build/configs/qemu */
patch -p1 < qemu-2.10.0-rc2_16m_ram_size.patch                                   
./configure --target-list=arm-softmmu                                            
make -j 4                                                                        
cd arm-softmmu/                                                                  
sudo ln qemu-system-arm /usr/local/bin/qemu-system-arm

Steps to run TizenRT:
qemu-system-arm -M lm3s6965evb -kernel build/output/bin/tinyara -nographic

Expected Result:
LM3S6965EVB : FLASH (B) = 134217728 : SRAM (B) = 16777216 : SDRAM (B) = 536870912
ABCDF
board_initialize: /dev/smart0p0 is mounted successfully at /mnt 
System Information:
	Version: 1.1
	Commit Hash: a764d1e00b0d2c1b98cc4e83b712f7423a680689
	Build User: manohara.hk@manohara-ubuntu
	Build Time: 2018-02-01 09:44:17
	System Time: 08 Jun 2010, 00:00:00 [s] UTC 
TASH>>Hello, World!!


SmartFS Test:
TASH>>smart_test -l 20 /mnt/file1.txt
Creating file /mnt/file1.txt for write mode
Writing test data.  20 lines to write
19
Done.
TASH>>cat /mnt/file1.txt
This is line 0 at offset 0
This is line 1 at offset 27
This is line 2 at offset 55
This is line 3 at offset 83
This is line 4 at offset 111
This is line 5 at offset 140
This is line 6 at offset 169
This is line 7 at offset 198
This is line 8 at offset 227
This is line 9 at offset 256
This is line 10 at offset 285
This is line 11 at offset 315
This is line 12 at offset 345
This is line 13 at offset 375
This is line 14 at offset 405
This is line 15 at offset 435
This is line 16 at offset 465
This is line 17 at offset 495
This is line 18 at offset 525
This is line 19 at offset 555

TASH>>

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by manoharahk** [comment]
Dear pillip8282 ,

Is it working for you now?

**Comment by pillip8282** [comment]
@ manoharahk
Sorry for the late reply.
I've tested it and it's working.

File: build/configs/qemu/qemu-2.10.0-rc2_16m_ram_size.patch [hunk 13]
```diff
@@ -1,7 +1,7 @@
-From a9f770243ccb354e99b9ff33ecfecbeca73b1b03 Mon Sep 17 00:00:00 2001
+From 23caf033c718dd932efd225a01d00587b88b081a Mon Sep 17 00:00:00 2001
```
Comment (Taejun-Kwon) [inline build/configs/qemu/qemu-2.10.0-rc2_16m_ram_size.patch]: Is this file needed??????

Comment (manoharahk) [reply build/configs/qemu/qemu-2.10.0-rc2_16m_ram_size.patch]: yes. Currently flash is made as readonly(ROM) inside qemu. 
Hence i have updated that patch to support flash as "read/write" inside qemu. 
Without that smartfs wont work

File: os/arch/arm/src/lm3s6965-ek/src/up_boot.c (hunk 12 above)
Comment (Taejun-Kwon) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: Now sflash use TIVA_FLASH_PAGESIZE * TIVA_FLASH_NPAGES = 256K
You set 1024 as a nblocks when call mtd_partition below.
This is something wrong.
To sum up,
1. Total Flash size 256*1024 = 256K
2. Erase Block size = 1024B
3. Number of Erase Block = 256
4. Page size = 1024B
But you defined number of page as a 1024
Please double check this.

Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: TIVA_FLASH_NPAGES is 128K not 256 when running under QEMU.

Please check the changes in " os/arch/arm/src/tiva/chip/lm3s_flash.h"

#if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM)                        
/* These parts all support a 1KiB erase page size and a total FLASH memory size  
 * of 128MB or 128K pages.                                                          
 */                                                                              
#define TIVA_FLASH_NPAGES        (128*1024)

Comment (Taejun-Kwon) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: 128 also problem, total size is TIVA_FLASH_NPAGES * TIVA_FLASH_PAGESIZE.
See ioctl, 
geo->blocksize & geo->erasesize is TIVA_FLASH_PAGESIZE,(Bytes)
geo->neraseblocks is TIVA_VIRTUAL_NPAGES
It means whole of flash is consist of 1024 * 128 = 128K

but you calling below
mtd_partition(mtd, partoffset, partsize / geo.blocksize, partno);
firstblock = (partoffset : 0)
nblocks = partsize/geo.blocksize : (1024 * 1024) / 1024

How allocate 1024K in 128K flash?

Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: Flash size is 128MB in case of QEMU.
Page size is 1024B, Hence TIVA_FLASH_NPAGES = 128 * 1024. =128K.

#define TIVA_VIRTUAL_NPAGES (TIVA_FLASH_NPAGES  - CONFIG_TIVA_FLASH_STARTPAGE).

Where,
CONFIG_TIVA_FLASH_STARTPAGE = 65536 (64K th page). 

Hence,
TIVA_VIRTUAL_NPAGES = 64K

So from 0 - 64MB of flash is not used . From 64MB offset, 1MB flash is used as a smartfs partition.

Comment (Taejun-Kwon) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: you don't understand what I mean..
If below selected by config value, mount will be failed.
#define TIVA_FLASH_NPAGES        256
I am saying that you had better use MTDIOC_GEOMETRY instead of define value, because NPAGES changes significantly based on config.

Comment (manoharahk) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: This code(mtd partition & mount)  is enabled only if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM).

And #if defined(CONFIG_QEMU_SRAM) || defined (CONFIG_QEMU_SDRAM),
then  #define TIVA_FLASH_NPAGES        (128*1024)

So How it will fail?

Comment (Taejun-Kwon) [reply os/arch/arm/src/lm3s6965-ek/src/up_boot.c]: ah.. ok then I didn't check this file fully, I didn't check whole code is in those configs.
Then It is acceptable.
